crates/forge_analyzer/src/interp.rs

```diff
@@ -676,9 +564,190 @@ impl<'cx, C: Runner<'cx>> Interp<'cx, C> {
+    #[inline]
+    pub fn add_value_to_definition(&mut self, defid_block: DefId, lval: Variable, rvalue: Rvalue) {
+        if let Variable {
+            base: Base::Var(varid),
+            projections,
+        } = lval
+        {
+            let (varid, projections) = self.get_farthest_obj(defid_block, varid, projections);
+            let rval_value = self.value_from_rval(defid_block, rvalue);
+            if let Some(existing_lval) = self
+                .get_value(defid_block, varid, Some(projections.clone()))
+                .cloned()
+            {
+                // if there is an existing value...
+                match (existing_lval, rval_value) {
+                    // return unknown if either values are unknown
+                    (Value::Unknown, _)
+                    | (_, Value::Unknown)
+                    | (Value::Const(_), Value::Object(_))
+                    | (Value::Phi(_), Value::Object(_))
+                    | (Value::Object(_), Value::Phi(_))
+                    | (Value::Object(_), Value::Const(_)) => self.add_value_with_projection(
+                        defid_block,
+                        varid,
+                        Value::Unknown,
+                        projections,
+                    ),
+                    // push other const onto phi vec if either are const and phi
+                    (Value::Const(const_value), Value::Phi(phi_value))
+                    | (Value::Phi(phi_value), Value::Const(const_value)) => {
+                        let mut new_phi = phi_value;
+                        new_phi.push(const_value);
+                        self.add_value_with_projection(
+                            defid_block,
+                            varid,
+                            Value::Phi(new_phi),
+                            projections,
+                        )
+                    }
+                    // push consts into vec if both are consts
+                    (Value::Const(const_value1), Value::Const(const_value2)) => self
+                        .add_value_with_projection(
+                            defid_block,
+                            varid,
+                            Value::Phi(vec![const_value1, const_value2]),
+                            projections,
+                        ),
+                    (Value::Object(exist_var), Value::Object(new_var)) => {
+                        // store projection values that are transferred
+                        let mut projections_transferred = vec![];
+                        // tranfer all projection values from the new_var into the existing var
+                        for ((_, _, projections), value) in
+                            self.value_manager.varid_to_value.clone().range(
+                                (defid_block, new_var, ProjectionVec::new())
+                                    ..(defid_block, VarId(new_var.0 + 1), ProjectionVec::new()),
```

Comment:
There's technically a bug in the event that `new_var.0 = u32::MAX`, since this would overflow back to 0. You can either create separate ranges for this case:

```rust
let start = (defid_block, new_var, smallvec![]);
let query = match new_var.0.checked_add(1) {
    Some(end) => start..(defid_block, VarId(end), smallvec![]),
    None => start..
};
for placeholder in self.value_manager.var_id_to_value.range(query) {
// etc
```

See [u32::checked_add](https://doc.rust-lang.org/std/primitive.u32.html#method.checked_add) for more details. Also, instead of cloning the entire `value_manager.varid_to_value`, you can instead create a `Vec` of the cloned `(value, projections)` using [Iterator::collect](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect) or `collect_vec`, and loop over them and call `self.add_value_with_projection` afterwards.

Reply:
I've tried googling to get the end if the `checked_add(1)` is none. Is there anyway to do this? I couldn't see anyway to do this....

Reply:
You use a [one-sided range](https://doc.rust-lang.org/std/ops/struct.RangeFrom.html), since if `checked_add(1)` is `None`, then you just need the rest of the map. The one-sided range is `start..` in this case.

Reply:
I am getting this compilation error. Do I need to find the max element of the var and the max element of the projection vec? Sorry for so many questions.... 

I was trying to find a way to solve it, but I was unsuccessful 😅 

```
25 |                           let query = match new_var.0.checked_add(1) {
    |  _____________________________________-
626 | |                             Some(end) => start..(defid_block, VarId(end), ProjectionVec::new()),
    | |                                          ------------------------------------------------------ this is found to be of type `std::ops::Range<(DefId, VarId, SmallVec<[Projection; 1]>)>`
627 | |                             None => start..
    | |                                     ^^^^^^^ expected `Range<(DefId, VarId, ...)>`, found `RangeFrom<(DefId, VarId, ...)>`
628 | |                         };
    | |_________________________- `match` arms have incompatible types
    |
```

Reply:
Ah, my bad. You need to put the queries in the different arms like so: 

```rust
let start = (defid_block, new_var, smallvec![]);
let query = match new_var.0.checked_add(1) {
    Some(end) => self.value_manager.var_id_to_value.range(start..(defid_block, VarId(end), smallvec![])),
    None => self.value_manager.var_id_to_value.range(start..),
};
for placeholder in query {
// etc
```

I forgot that `x..y` and `x..` are two different types.